### ISSUE TITLE
The forward image features of samples for models with and without BetterTransformer transformation should be nearly identical, but are in practice often very different.

### ISSUE THREAD

**Issue by jyangliu** [huggingface/optimum#1437]
### System Info

```shell
optimum:1.13.2
transformers:4.33.3
win10/CUDA 11.7/pytorch 2.0.0
```

### Who can help?

@philschmid @fxmarty @younesbelkada

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

[https://github.com/jyangliu/bettertransformer_test/blob/main/test_forward.py](https://github.com/jyangliu/bettertransformer_test/blob/main/test_forward.py)

### Expected behavior

The forward image features of samples for models with and without BetterTransformer transformation should be nearly identical, but are in practice often very different.

**Comment by jyangliu** [comment]
I find that optimum==1.9.0 can work correctly, but can not support clip trainning in optimum==1.9.0. Optimum==1.13.2  support clip training, but got wrong forward features.

**Comment by fxmarty** [comment]
Thank you I will have a look at it. Our CI is green but probably not strong enough.

**Comment by fxmarty** [comment]
If you have a reproducible example that would help!

**Comment by jyangliu** [comment]
Please check code in here [https://github.com/jyangliu/bettertransformer_test/blob/main/test_forward.py](https://github.com/jyangliu/bettertransformer_test/blob/main/test_forward.py)

**Comment by fxmarty** [comment]
Thank you @jyangliu, interestingly I can reproduce the issue only on cuda device, not cpu.

First, you probably see the warnings `Overridding quick_gelu activation with gelu. Use the transformed model at your own risk, the output logits could be significantly different.`. The implementation in BetterTransformer only matches with RELU and GELU, so some difference is expected vs quick gelu.

Printing the average relative difference, it is about ~2 % due to this.

Then, just for the purpose of it, replacing quick gelu by gelu, the output indeed matches on CPU device but not on CUDA device:

```python
import torch
from optimum.bettertransformer import BetterTransformer

from transformers import CLIPModel

import copy
from transformers.activations import GELUActivation

model = CLIPModel.from_pretrained("openai/clip-vit-base-patch16")
model.eval()
device = "cuda"

for i in range(len(model.vision_model.encoder.layers)):
    model.vision_model.encoder.layers[i].mlp.activation_fn = GELUActivation()

pixel_values = torch.randn(1, 3, 448//2, 448//2).float()
pixel_values1 = copy.deepcopy(pixel_values)

pixel_values = pixel_values.to(device)
pixel_values1 = pixel_values1.to(device)

model = model.to(device)
with torch.no_grad():
    y1 = model.get_image_features(pixel_values=pixel_values)

model_bt = BetterTransformer.transform(model)

model_bt = model_bt.to(device)

with torch.no_grad():
    y2 = model_bt.get_image_features(pixel_values=pixel_values)

    print("mean absdiff", torch.mean((y1 - y2) / (y1 + 1e-12)))
    assert torch.allclose(y1, y2, atol=1e-3)
```

I am still getting a mean absolute difference of ~0.1 - 0.5%.

It is easy to check that the issue is coming from PyTorch:

```python
hidden_states = torch.rand(1, 197, 768).to(device)

res_cuda = model.vision_model.encoder.layers[i](hidden_states, attention_mask=None, causal_attention_mask=None, output_attentions=False)[0]

cpu_layer = model.vision_model.encoder.layers[i].to("cpu")
res_cpu = cpu_layer(hidden_states.to("cpu"), attention_mask=None, causal_attention_mask=None, output_attentions=False)[0]

res_cuda = res_cuda.cpu()
print("mean absdiff", torch.mean((res_cpu - res_cuda) / (res_cpu + 1e-12)).abs())
```

**Comment by jyangliu** [comment]
Thank for you reply @fxmarty . As you said, It can work with model.eval() mode. But when I change to model.train() model, it get wrong result. I check the CLIP code, and find no BN or Dropout use in the code. Could you please have a look for this?

`import torch
from optimum.bettertransformer import BetterTransformer

from transformers import CLIPModel

import copy
from transformers.activations import GELUActivation

model = CLIPModel.from_pretrained("openai/clip-vit-base-patch16")
model.train()
device = "cuda"

for i in range(len(model.vision_model.encoder.layers)):
    model.vision_model.encoder.layers[i].mlp.activation_fn = GELUActivation()

pixel_values = torch.randn(1, 3, 448//2, 448//2).float()
pixel_values1 = copy.deepcopy(pixel_values)

pixel_values = pixel_values.to(device)
pixel_values1 = pixel_values1.to(device)

model = model.to(device)
with torch.no_grad():
    y1 = model.get_image_features(pixel_values=pixel_values)

model_bt = BetterTransformer.transform(model)

model_bt = model_bt.to(device)

with torch.no_grad():
    y2 = model_bt.get_image_features(pixel_values=pixel_values)

    print("mean absdiff", torch.mean((y1 - y2) / (y1 + 1e-12)))
    assert torch.allclose(y1, y2, atol=1e-3)`

**Comment by jyangliu** [comment]
Thank for you reply @fxmarty . As you said, It can work with model.eval() mode. But when I change to model.train() model, it get wrong result. I check the CLIP code, and find no BN or Dropout use in the code. Could you please have a look for this?

`import torch
from optimum.bettertransformer import BetterTransformer

from transformers import CLIPModel

import copy
from transformers.activations import GELUActivation

model = CLIPModel.from_pretrained("openai/clip-vit-base-patch16")
model.train()
device = "cuda"

for i in range(len(model.vision_model.encoder.layers)):
    model.vision_model.encoder.layers[i].mlp.activation_fn = GELUActivation()

pixel_values = torch.randn(1, 3, 448//2, 448//2).float()
pixel_values1 = copy.deepcopy(pixel_values)

pixel_values = pixel_values.to(device)
pixel_values1 = pixel_values1.to(device)

model = model.to(device)
with torch.no_grad():
    y1 = model.get_image_features(pixel_values=pixel_values)

model_bt = BetterTransformer.transform(model)

model_bt = model_bt.to(device)

with torch.no_grad():
    y2 = model_bt.get_image_features(pixel_values=pixel_values)

    print("mean absdiff", torch.mean((y1 - y2) / (y1 + 1e-12)))
    assert torch.allclose(y1, y2, atol=1e-3)`

**Comment by jyangliu** [comment]
Thank for you reply @fxmarty . As you said, It can work with model.eval() mode. But when I change to model.train() model, it get wrong result. I check the CLIP code, and find no BN or Dropout use in the code. Could you please have a look for this?

**Comment by fxmarty** [comment]
Thank you @jyangliu, could you provide a reproduction?

I don't expect CLIP BetterTransformer to work in training mode, it is not implemented: https://github.com/huggingface/optimum/blob/e7bd60dd2c1e295263ba57a4e468a62ab5b179e8/optimum/bettertransformer/models/encoder_models.py#L1807

**Comment by jyangliu** [comment]
I only change model.eval() to model.train(), and it can still run. 
```
import torch
from optimum.bettertransformer import BetterTransformer

from transformers import CLIPModel

import copy
from transformers.activations import GELUActivation

model = CLIPModel.from_pretrained("openai/clip-vit-base-patch16")
model.train()
device = "cuda"

for i in range(len(model.vision_model.encoder.layers)):
    model.vision_model.encoder.layers[i].mlp.activation_fn = GELUActivation()

pixel_values = torch.randn(1, 3, 448//2, 448//2).float()
pixel_values1 = copy.deepcopy(pixel_values)

pixel_values = pixel_values.to(device)
pixel_values1 = pixel_values1.to(device)

model = model.to(device)
with torch.no_grad():
    y1 = model.get_image_features(pixel_values=pixel_values)

model_bt = BetterTransformer.transform(model)

model_bt = model_bt.to(device)

with torch.no_grad():
    y2 = model_bt.get_image_features(pixel_values=pixel_values)

    print("mean absdiff", torch.mean((y1 - y2) / (y1 + 1e-12)))
    assert torch.allclose(y1, y2, atol=1e-3)
```

Meanwhile, Is clip bettertransformer training mode implemented for optimum==1.13.2? It can run in training mode without any implemenet error when I use backward() function.

```
import torch
from optimum.bettertransformer import BetterTransformer

from transformers import CLIPModel

import copy
from transformers.activations import GELUActivation
from torch.autograd import Variable

model = CLIPModel.from_pretrained("openai/clip-vit-base-patch16")
model.train()
device = "cuda"

for i in range(len(model.vision_model.encoder.layers)):
    model.vision_model.encoder.layers[i].mlp.activation_fn = GELUActivation()

pixel_values = Variable(torch.randn(1, 3, 448//2, 448//2).float())
pixel_values1 = copy.deepcopy(pixel_values)

pixel_values = pixel_values.to(device)
pixel_values1 = pixel_values1.to(device)

model = model.to(device)
y1 = model.get_image_features(pixel_values=pixel_values)

model_bt = BetterTransformer.transform(model)

model_bt = model_bt.to(device)

y2 = model_bt.get_image_features(pixel_values=pixel_values)
loss = y2.mean()
loss.backward()

print("mean absdiff", torch.mean((y1 - y2) / (y1 + 1e-12)))
assert torch.allclose(y1, y2, atol=1e-3)
```

**Comment by fxmarty** [comment]
@jyangliu thank you for the notice, indeed an error should be raised but was not. Here is the fix: https://github.com/huggingface/optimum/pull/1468
